queue.go

```diff
@@ -395,11 +395,19 @@ func (queue *redisQueue) consumerBatchConsume(batchSize int, timeout time.Durati
 		// debug(fmt.Sprintf("batch consume consume %d", len(batch))) // COMMENTOUT
 		consumer.Consume(batch)
 
-		// reset batch and timer
-		batch = batch[:0]
-		if !timer.Stop() {
-			<-timer.C
-		}
+		batch = batch[:0] // reset batch
+		stopTimer(timer)  // stop and drain the timer if it fired in between
+	}
+}
+
+func stopTimer(timer *time.Timer) {
```

Comment:
I am happy with this. I suspect that there is still a race condition here where time.Stop() returns false - but the timer.C hasn't quite gotten a value in it and the select misses that. This will produce a spurious timer firing.

In this case you will just break the loop early and probably find and empty batch slice. I think this will be very rare, if it happens at all, and doesn't cause any harm.

I'll have more of a play around with timers to see what the heck is going on with them.